cibuildwheel/linux.py

```diff
@@ -111,24 +112,37 @@ def build(options: BuildOptions) -> None:
     container_package_dir = container_project_path / abs_package_dir.relative_to(cwd)
     container_output_dir = PurePath('/output')
 
+    log = Logger()
+
     for implementation, platform_tag, docker_image in platforms:
         platform_configs = [c for c in python_configurations if c.identifier.startswith(implementation) and c.identifier.endswith(platform_tag)]
         if not platform_configs:
             continue
 
         try:
+            log.step(f'Starting Docker image {docker_image}...')
             with DockerContainer(docker_image, simulate_32_bit=platform_tag.endswith('i686'), cwd=container_project_path) as docker:
+                log.step_end()
+
+                log.step('Copying project into Docker...')
```

Comment:
It would be great to also be able to use a context manager no ? It clearly shows the intent and prevents forgetting to call `log.step_end()` at the end of that intent (even though I'm seeing now that steps are closed automatically when a new one starts)
e.g.

```
with log.step('Copying project into Docker...'):
    docker.copy_into(Path.cwd(), container_project_path)
```
even more for longer process:
```
with log.step('Running before_all...'):
    env = docker.get_environment()
    env['PATH'] = f'/opt/python/cp38-cp38:{env["PATH"]}'
    env = options.environment.as_dictionary(env, executor=docker.environment_executor)

    before_all_prepared = prepare_command(options.before_all, project=container_project_path, package=container_package_dir)
    docker.call(['sh', '-c', before_all_prepared], env=env)
```

Reply:
I originally built the whole thing using context managers in fact, you can see that here:
https://github.com/joerick/cibuildwheel/blob/946306b21dee8411cc27807a9f47a7e91ecf1b5d/cibuildwheel/linux.py

Reply:
[whoops, hit submit somehow!]

I concluded that adding an extra two levels of indentation to an already very indented function was too much, and yes, I couldn't wrap the Docker start, so split it into a `build_one` function, at which point it felt like the tail wagging the dog - the logging functionality is mandating the program structure? So I decided to switch to the 'imperative' style as seen in https://github.com/joerick/cibuildwheel/commit/38e7b5a91ded109ae44947b754cd47d28e3a26d5 .

This seemed a little more lightweight.

But I also don't really like those `step_end()` method calls. One option (I think @mayeut suggests) is that I return a context manager from step(), whose `__exit__` calls step_end, but it's usage is optional. I might give that a go to see how it feels.

Reply:
> Why do we need the succeeding step_end, though, if it's automatically closed anyway?

Also a good point, I could remove that one.

Reply:
> I concluded that adding an extra two levels of indentation to an already very indented function was too much

Yeah, now that you're saying this, I can see that you're right.

Reply:
The nice thing about a context manager is that it always closes, even if an error is thrown. It seems to me there's at least one path with where the close may not happen - I assume it's not critical, but maybe you could have the first step be a context manager, and remove the final end. Then there would just be a `.step` which would close and start a new step. You can use `contextlib.ExitStack` to keep from indenting too much. 

I'm a huge fan of context managers; I added finally context managers to my LCD code in my vent monitor project and suddenly a long standing bug went away - there was a way the Rπ's resources were not being cleaned up, and now they were.

This looks good as is, though, just a general suggestion.